### PR TITLE
Fix scores not being ordered correctly on leaderboards

### PR DESCRIPTION
**Pull request by smoogipoo**
Prereqs:
- [x] https://github.com/ppy/osu/pull/14598

Fixes the order portion of https://github.com/ppy/osu/issues/12305. Accuracy is TBD in another PR.

Because calculating total score can depend on difficulty calculation in the case of legacy scores, I've gone through usage scenarios to make sure that the diffcalc portion of it is fully backgrounded and doesn't cause a large stutter. A large part of this is working around the fact that we don't yet have update-thread synchronization contexts, but I don't think it looks too bad.

### CODE REVIEW

**Comment by peppy** [comment]
On an initial pass this all looks pretty sane flow-wise. Will wait for test passingness and do some final checks.

**Comment by peppy** [comment]
@smoogipoo just a prod that this is still failing

**Comment by smoogipoo** [comment]
Tests should be fixed now.

File: osu.Game.Tests/Visual/UserInterface/TestSceneDeleteLocalScore.cs [hunk 1]
```diff
@@ -158,14 +158,14 @@ public void TestDeleteViaRightClick()
                 InputManager.Click(MouseButton.Left);
             });
 
-            AddUntilStep("score removed from leaderboard", () => leaderboard.Scores.All(s => s.OnlineScoreID != scores[0].OnlineScoreID));
+            AddUntilStep("score removed from leaderboard", () => leaderboard.Scores.All(s => s != scores[0]));
```
Comment (peppy) [inline osu.Game.Tests/Visual/UserInterface/TestSceneDeleteLocalScore.cs:161]: just out of curiosity, why was this required?

Comment (smoogipoo) [reply osu.Game.Tests/Visual/UserInterface/TestSceneDeleteLocalScore.cs:161]: Because they're sorted by total score (`ScoreInfo.TotalScore`).

Comment (peppy) [reply osu.Game.Tests/Visual/UserInterface/TestSceneDeleteLocalScore.cs:161]: Right.. that's a weird test scene structure. Might change it while we're here.

File: osu.Game/Screens/Ranking/ScorePanelList.cs [hunk 2]
```diff
@@ -114,36 +130,56 @@ public ScorePanel AddScore(ScoreInfo score, bool isNewLocalScore = false)
...
 
+        private void displayScore(ScorePanelTrackingContainer trackingContainer)
+        {
+            if (!IsLoaded)
+                return;
+
+            var score = trackingContainer.Panel.Score;
+
+            // Calculating score can take a while in extreme scenarios, so only display scores after the process completes.
+            scoreManager.GetTotalScoreAsync(score)
```
Comment (peppy) [inline osu.Game/Screens/Ranking/ScorePanelList.cs:156]: Just on an initial read, does the fact that this is done at a per panel level mean that the list could be reordered multiple times during a single update pass (ie. if the user changes the score display mode)?

**Comment by peppy** [comment]
I noticed that the scores seem to be flashing on `BeatmapLeaderboard` when changing the scoring mode, but this does look to also be occurring on `master`:

https://user-images.githubusercontent.com/191335/132315824-05f9152d-2ac3-414d-bcdc-07e6252c4c60.mp4

And in trying to look further into it, couldn't find a test where I can toggle the classic/standardised scoring display on `BeatmapLeaderboard` for testing. Is there a scene I'm missing which covers this?

**Comment by smoogipoo** [comment]
I don't know. My changes don't touch that part of code.

**Comment by peppy** [comment]
Yeah, for sure. What I was getting at is that I think we probably want test coverage of that flow though? Right now I can't find a test scene which actually runs that fetch-then-order-by (have been testing via the game itself).

**Comment by smoogipoo** [comment]
Sure? It's orthogonal to this PR though?

**Comment by peppy** [comment]
Hmm, I guess? Wanted to make sure we had test coverage of the initial sort. Guess it wasn't tested before and still isn't tested so nothing lost/gained..

File: osu.Game.Tests/Visual/UserInterface/TestSceneDeleteLocalScore.cs (hunk 1 above)
Comment (peppy) [reply osu.Game.Tests/Visual/UserInterface/TestSceneDeleteLocalScore.cs]: Has been fixed in https://github.com/ppy/osu/pull/14603/commits/d922210d2feda6f03de5717d3ca53e99b4864feb (was an actual issue, as discussed on discord).

File: osu.Game/Screens/Ranking/ScorePanelList.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Screens/Ranking/ScorePanelList.cs:156]: I guess that's just this method name being misleading. Okay to rename this method to `focusScore`?

Comment (smoogipoo) [reply osu.Game/Screens/Ranking/ScorePanelList.cs:156]: Score display mode should not/does not reorder scores. That was the reason for https://github.com/ppy/osu/pull/14598.

But if the order of scores has changed when compared to osu!stable, then yes, you will get panels inserted inbetween other panels. Fixing that would require the API to paginate by the standardised total score instead.

Comment (peppy) [reply osu.Game/Screens/Ranking/ScorePanelList.cs:156]: In practice this looks to work pretty well, so gonna resolve this for now.

File: osu.Game/Screens/Ranking/ResultsScreen.cs [hunk 3]
```diff
@@ -191,19 +190,18 @@ protected override void Update()
...
+                APIRequest nextPageRequest = null;
+
                 if (ScorePanelList.IsScrolledToStart)
                     nextPageRequest = FetchNextPage(-1, fetchScoresCallback);
                 else if (ScorePanelList.IsScrolledToEnd)
                     nextPageRequest = FetchNextPage(1, fetchScoresCallback);
 
                 if (nextPageRequest != null)
                 {
-                    // Scheduled after children to give the list a chance to update its scroll position and not potentially trigger a second request too early.
```
Comment (peppy) [inline osu.Game/Screens/Ranking/ResultsScreen.cs:203]: Did you confirm that this reasoning is no longer a requirement?

Comment (smoogipoo) [reply osu.Game/Screens/Ranking/ResultsScreen.cs:203]: This entire code has changed. Yes, I've confirmed it. There's tests for this as well (`TestScenePlaylistsResultsScreen` and the ones which test scroll to left/scroll to right with delayed load).

File: osu.Game/Scoring/ScoreManager.cs [hunk 4]
```diff
@@ -103,6 +106,37 @@ protected override bool CheckLocalAvailability(ScoreInfo model, IQueryable<Score
             => base.CheckLocalAvailability(model, items)
                || (model.OnlineScoreID != null && items.Any(i => i.OnlineScoreID == model.OnlineScoreID));
 
+        /// <summary>
+        /// Orders an array of <see cref="ScoreInfo"/>s by total score.
+        /// </summary>
+        /// <param name="scores">The array of <see cref="ScoreInfo"/>s to reorder.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> to cancel the process.</param>
+        /// <returns>The given <paramref name="scores"/> ordered by decreasing total score.</returns>
+        public async Task<ScoreInfo[]> OrderByTotalScoreAsync(ScoreInfo[] scores, CancellationToken cancellationToken = default)
```
Comment (peppy) [inline osu.Game/Scoring/ScoreManager.cs:115]: Are you okay with this structure instead? Seems more logical:

```csharp
        public async Task<ScoreInfo[]> OrderByTotalScoreAsync(ScoreInfo[] scores, CancellationToken cancellationToken = default)
        {
            var difficultyCache = difficulties?.Invoke();

            if (difficultyCache != null)
            {
                // Compute difficulties asynchronously first to prevent blocking via the GetTotalScore() call below.
                foreach (var s in scores)
                {
                    await difficultyCache.GetDifficultyAsync(s.Beatmap, s.Ruleset, s.Mods, cancellationToken).ConfigureAwait(false);
                    cancellationToken.ThrowIfCancellationRequested();
                }
            }

            // We're calling .Result, but this should not be a blocking call due to the above GetDifficultyAsync() calls.
            return scores.OrderByDescending(s => GetTotalScoreAsync(s, cancellationToken: cancellationToken).Result)
                         .ThenBy(s => s.OnlineScoreID)
                         .ToArray();
        }
```

Comment (smoogipoo) [reply osu.Game/Scoring/ScoreManager.cs:115]: That's fine.

**Review by peppy — APPROVED** [review]
